### PR TITLE
[ADP-2959] Simplify delegation and payment address classes

### PR DESCRIPTION
**Pull request by paolino**
- [x] Remove `n` parameter from `PaymentAddress` and `DelegationAddress` classes and move it to the methods as singleton `SNetworkId n` parameter
- [x] Add some functions that simplify the call of the methods in the old way (forcing `@n`) so as not to change too much the call site
- [x]  Remove `:: NetworkDiscriminant` superfluous signature around
- [x] Remove redundant constraints on `PaymentAddress` and `DelegationAddress` now that `n` is removed

ADP-2959

### CODE REVIEW

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server.hs [hunk 1]
```diff
@@ -319,7 +314,7 @@ server byron icarus shelley multisig spl ntp blockchainSource =
     coinSelections :: Server (CoinSelections n)
     coinSelections = (\wid -> \case
         ApiSelectForPayment ascp ->
-            selectCoins shelley (delegationAddress @n) wid ascp
+            selectCoins shelley (delegationAddressS @n) wid ascp
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Server.hs:317]: At some point, we probably want to avoid the ambiguous function `delegationAddressS` (which requires a type signature `@n`), and write something along the lines of `delegationAddress (sNetworkId @n)`, but I agree that it's better to leave that to another pull request. For now, removing the class constraints has priority. 👍

File: lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs [hunk 2]
```diff
@@ -287,70 +289,63 @@ instance WalletKey ShelleyKey where
 instance GetPurpose ShelleyKey where
     getPurpose = purposeCIP1852
 
-instance PaymentAddress 'Mainnet ShelleyKey 'CredFromKeyK where
-    paymentAddress paymentK = do
+instance PaymentAddress ShelleyKey 'CredFromKeyK where
+    paymentAddress SMainnet paymentK = do
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs:293]: The code is as correct as the previous version. 👍  The diff is a big mangled due to duplication, though.

Could you do a follow-up pull request where you group the case distinction on the networkId into a single definition of `networkTag`? I.e.

```hs
    paymentAddress SMainnet paymentK = do
        Address $ BL.toStrict $ runPut $ do
            putWord8 (enterprise + networkTag)
            putByteString . blake2b224 . xpubPublicKey . getKey $ paymentK
      where
        enterprise = 96
        networkTag = case networkId of
            SMainnet -> 1
            STestnet _ -> 0
```

(To reduce code duplication, it would also make sense to define `paymentAddress` in terms of `liftPaymentAddress`, but let's not go overboard — at some point, we may want to switch to `cardano-addresses` instead.)

File: lib/wallet/src/Cardano/Wallet/Primitive/Types/ProtocolMagic.hs [hunk 3]
```diff
@@ -56,3 +60,7 @@ mainnetMagic =  ProtocolMagic 764824073
 -- | Derive testnet magic from a type-level Nat
 testnetMagic :: forall pm. KnownNat pm => ProtocolMagic
 testnetMagic = ProtocolMagic $ fromIntegral $ natVal $ Proxy @pm
+
+magicSNetworkId :: SNetworkId n -> ProtocolMagic
+magicSNetworkId SMainnet = ProtocolMagic  764824073
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Primitive/Types/ProtocolMagic.hs:65]: 'tis magic, hardcoded! 🧚

File: lib/wallet/src/Cardano/Wallet/Shelley/Network/Discriminant.hs [hunk 4]
```diff
@@ -50,14 +42,8 @@ import qualified Cardano.Ledger.BaseTypes as Ledger
...
 data SomeNetworkDiscriminant where
     SomeNetworkDiscriminant
-        :: forall (n :: NetworkDiscriminant).
-            ( PaymentAddress n IcarusKey 'CredFromKeyK
-            , PaymentAddress n ByronKey 'CredFromKeyK
-            , PaymentAddress n ShelleyKey 'CredFromKeyK
-            , DelegationAddress n ShelleyKey 'CredFromKeyK
-            , HasSNetworkId n
-            , Typeable n
-            )
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Shelley/Network/Discriminant.hs:60]: 🤤

**Review by HeinrichApfelmus — APPROVED** [review]
Excellent, thank you! 😊 Happy to merge, but request for a small follow-up pull request.

The introduction of e.g. `liftDelegationAddressS` is a bit unfortunate, but makes the pull request easier to review. 👍

File: lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs (hunk 2 above)
Comment (HeinrichApfelmus) [reply lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs:293]: (Same for the other classes e.g. `delegationAddress`)

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server.hs (hunk 1 above)
Comment (paolino) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Server.hs:317]: yup, that was the intention as you can already appreciate the impact of this PR :-/

File: lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs (hunk 2 above)
Comment (paolino) [reply lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs:293]: yes

File: lib/wallet/src/Cardano/Wallet/Primitive/Types/ProtocolMagic.hs (hunk 3 above)
Comment (paolino) [reply lib/wallet/src/Cardano/Wallet/Primitive/Types/ProtocolMagic.hs:65]: someone's telephone number

**Comment by paolino** [comment]
bors ping
